engine/consensus/matching/engine.go

```diff
@@ -651,37 +703,75 @@ func (e *Engine) sealableResults() ([]*flow.IncorporatedResult, error) {
 		// matching engine must enforce equality of start and end state for a result with zero chunks,
 		// in the absence of anyone else doing do.
 		matched := false
+		unmatchedIndex := -1
 		// check that each chunk collects enough approvals
-		for _, chunk := range incorporatedResult.Result.Chunks {
+		for i, chunk := range incorporatedResult.Result.Chunks {
 			matched, err = e.matchChunk(incorporatedResult, block, chunk, assignment)
 			if err != nil {
-				return nil, fmt.Errorf("could not match chunk: %w", err)
+				return nil, nil, fmt.Errorf("could not match chunk: %w", err)
 			}
 			if !matched {
+				unmatchedIndex = i
 				break
 			}
 		}
 
+		matchedAfterConsideringEmergency := matched
```

Comment:
To me it would be clearer to keep emergency sealing conceptually separate for now:

```suggestion
		emergencySealed := false
```

Then, the following logic would me more readable for me: 
* https://github.com/onflow/flow-go/blob/a412a0e844aff8adff1fa3a624e7fb8a3d74fc64/engine/consensus/matching/engine.go#L738-L741 would turn to
    ```golang 
    if matched || emergencySealed {
	    // add the result to the results that should be sealed
	    results = append(results, incorporatedResult)
    }
    ```
* https://github.com/onflow/flow-go/blob/a412a0e844aff8adff1fa3a624e7fb8a3d74fc64/engine/consensus/matching/engine.go#L755-L756 would turn into
    ```golang 
    CanBeSealed:              matched || emergencySealed,
    SealedByEmergency:        emergencySealed,
    ```